Quantum-Control-Applications/Superconducting/Single Flux Tunable Transmon/SSB mixer for readout/macros.py

```diff
@@ -72,49 +90,47 @@ def active_reset_until_success(threshold, max_tries=1, Ig=None):
-def ge_singleshot_measurement(cooldown_time):
-    """Macro measuring the singleshot qubit's ground and excited states.
-
-    :param cooldown_time: cooldown time between two successive qubit state measurements in clock cycle unit (4ns).
-    :return: singleshot I and Q data for the ground and excited states respectively [Ig, Qg, Ie, Qe].
+# Macro for measuring the qubit state with single shot
+def singleshot_measurement(threshold=None, state=None, I=None, Q=None):
     """
-    Ig = declare(fixed)
-    Qg = declare(fixed)
-    Ie = declare(fixed)
-    Qe = declare(fixed)
-
-    # Ground state measurement
-    align("qubit", "resonator")
-    measure(
-        "readout",
-        "resonator",
-        None,
-        dual_demod.full("rotated_cos", "out1", "rotated_sin", "out2", Ig),
-        dual_demod.full("rotated_minus_sin", "out1", "rotated_cos", "out2", Qg),
-    )
-    wait(cooldown_time, "resonator", "qubit")
-
-    # Excited state measurement
-    align("qubit", "resonator")
-    play("pi", "qubit")
-    align("qubit", "resonator")
+    A macro for performing the singleshot readout, with the ability to perform state discrimination.
+    If `threshold` is given, the information in the `I` quadrature will be compared against the threshold and `state`
+    would be `True` if `I > threshold`.
+    Note that it is assumed that the results are rotated such that all the information is in the `I` quadrature.
+
+    :param threshold: Optional. The threshold to compare `I` against.
+    :param state: A QUA variable for the state information, only used when a threshold is given.
+        Should be of type `bool`. If not given, a new variable will be created
+    :param I: A QUA variable for the information in the `I` quadrature. Should be of type `Fixed`. If not given, a new
+        variable will be created
+    :param Q: A QUA variable for the information in the `Q` quadrature. Should be of type `Fixed`. If not given, a new
+        variable will be created
+    :return: Three QUA variables populated with the results of the readout: (`state` (only if threshold is not None), `I`, `Q`)
+    """
+    if I is None:
+        I = declare(fixed)
+    if Q is None:
+        Q = declare(fixed)
+    if (threshold is not None) and (state is None):
+        state = declare(bool)
     measure(
         "readout",
         "resonator",
         None,
-        dual_demod.full("rotated_cos", "out1", "rotated_sin", "out2", Ie),
-        dual_demod.full("rotated_minus_sin", "out1", "rotated_cos", "out2", Qe),
+        dual_demod.full("rotated_cos", "out1", "rotated_sin", "out2", I),
+        dual_demod.full("rotated_minus_sin", "out1", "rotated_cos", "out2", Q),
     )
-    wait(cooldown_time, "resonator", "qubit")
-
-    return Ig, Qg, Ie, Qe
+    if threshold is not None:
+        assign(state, I > threshold)
+        return state, I, Q
+    else:
+        return I, Q
 
 
 # Macro for measuring the averaged ground and excited states for calibration
```

Comment:
is this being used?

Reply:
Yes because it allows to calibrate the I Q values for ground and excited in the case where you don't have a good single-shot fidelity to compute the population in the excited state. It is used for the cryoscope codes

Reply:
Ok, cool :)